specs/sharding/beacon-chain.md

```diff
@@ -103,786 +87,558 @@ The following values are (non-configurable) constants used throughout the specif
+    assert block.slot > state.latest_block_header.slot
+    # Verify that proposer index is the correct index
+    if not is_intermediate_block_slot(block.slot):
+        assert block.proposer_index == get_beacon_proposer_index(state)
+    # Verify that the parent matches
+    assert block.parent_root == hash_tree_root(state.latest_block_header)
+    # Cache current block as the new latest block
+    state.latest_block_header = BeaconBlockHeader(
+        slot=block.slot,
+        proposer_index=block.proposer_index,
+        parent_root=block.parent_root,
+        state_root=Bytes32(),  # Overwritten in the next process_slot call
+        body_root=hash_tree_root(block.body),
+    )
 
-```python
-def process_operations(state: BeaconState, body: BeaconBlockBody) -> None:
-    # Verify that outstanding deposits are processed up to the maximum number of deposits
-    assert len(body.deposits) == min(MAX_DEPOSITS, state.eth1_data.deposit_count - state.eth1_deposit_index)
-
-    def for_ops(operations: Sequence[Any], fn: Callable[[BeaconState, Any], None]) -> None:
-        for operation in operations:
-            fn(state, operation)
-
-    for_ops(body.proposer_slashings, process_proposer_slashing)
-    for_ops(body.attester_slashings, process_attester_slashing)
-    # New shard proposer slashing processing
-    for_ops(body.shard_proposer_slashings, process_shard_proposer_slashing)
-
-    # Limit is dynamic: based on active shard count
-    assert len(body.shard_headers) <= MAX_SHARD_HEADERS_PER_SHARD * get_active_shard_count(state, get_current_epoch(state))
-    for_ops(body.shard_headers, process_shard_header)
-
-    # New attestation processing
-    for_ops(body.attestations, process_attestation)
-    for_ops(body.deposits, process_deposit)
-    for_ops(body.voluntary_exits, process_voluntary_exit)
-
-    # TODO: to avoid parallel shards racing, and avoid inclusion-order problems,
-    #  update the fee price per slot, instead of per header.
-    # state.shard_sample_price = compute_updated_sample_price(state.shard_sample_price, ?, shard_count)
+    # Verify proposer is not slashed
+    proposer = state.validators[block.proposer_index]
+    assert not proposer.slashed
 ```
 
-##### Extended Attestation processing
+#### Intermediate Block Bid
 
 ```python
-def process_attestation(state: BeaconState, attestation: Attestation) -> None:
-    altair.process_attestation(state, attestation)
-    process_attested_shard_work(state, attestation)
-```
+def verify_intermediate_block_bid(state: BeaconState, block: BeaconBlock) -> None:
+    if is_intermediate_block_slot(block.slot):
+        # Get last intermediate block bid
+        assert state.blocks_since_intermediate_block[-1].body.intermediate_block_bid_with_recipient_address.selector == 1
+        intermediate_block_bid = state.blocks_since_intermediate_block[-1].body.intermediate_block_bid_with_recipient_address.value.intermediate_block_bid
```

Comment:
Hmm. `IntermediateBlockBidWithRecipientAddress.intermediate_block_bid` is also a `Union` but is not treated as such here and in the `else` clause.